meta/meta.go

```diff
@@ -679,22 +700,70 @@ func (m *Meta) FinishBootstrap(version int64) error {
 }
 
-// UpdateDDLReorgHandle saves the job reorganization latest processed handle for later resuming.
-func (m *Meta) UpdateDDLReorgHandle(job *model.Job, handle int64) error {
-	err := m.txn.HSet(mDDLJobReorgKey, m.jobIDKey(job.ID), []byte(strconv.FormatInt(handle, 10)))
+// UpdateDDLReorgStartHandle saves the job reorganization latest processed start handle for later resuming.
+func (m *Meta) UpdateDDLReorgStartHandle(job *model.Job, startHandle int64) error {
+	err := m.txn.HSet(mDDLJobReorgKey, m.reorgJobStartHandle(job.ID), []byte(strconv.FormatInt(startHandle, 10)))
+	return errors.Trace(err)
+}
+
+// UpdateDDLReorgHandle saves the job reorganization latest processed information for later resuming.
+func (m *Meta) UpdateDDLReorgHandle(job *model.Job, startHandle, endHandle, partitionID int64) error {
+	err := m.txn.HSet(mDDLJobReorgKey, m.reorgJobStartHandle(job.ID), []byte(strconv.FormatInt(startHandle, 10)))
+	if err != nil {
+		return errors.Trace(err)
+	}
+	err = m.txn.HSet(mDDLJobReorgKey, m.reorgJobEndHandle(job.ID), []byte(strconv.FormatInt(endHandle, 10)))
+	if err != nil {
+		return errors.Trace(err)
+	}
+	err = m.txn.HSet(mDDLJobReorgKey, m.reorgJobPartitionID(job.ID), []byte(strconv.FormatInt(partitionID, 10)))
 	return errors.Trace(err)
 }
 
 // RemoveDDLReorgHandle removes the job reorganization handle.
 func (m *Meta) RemoveDDLReorgHandle(job *model.Job) error {
-	err := m.txn.HDel(mDDLJobReorgKey, m.jobIDKey(job.ID))
-	return errors.Trace(err)
+	err := m.txn.HDel(mDDLJobReorgKey, m.reorgJobStartHandle(job.ID))
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if err = m.txn.HDel(mDDLJobReorgKey, m.reorgJobEndHandle(job.ID)); err != nil {
+		log.Warn(err)
+	}
+	if err = m.txn.HDel(mDDLJobReorgKey, m.reorgJobPartitionID(job.ID)); err != nil {
+		log.Warn(err)
+	}
+	return nil
 }
 
-// GetDDLReorgHandle gets the latest processed handle.
-func (m *Meta) GetDDLReorgHandle(job *model.Job) (int64, error) {
-	value, err := m.txn.HGetInt64(mDDLJobReorgKey, m.jobIDKey(job.ID))
-	return value, errors.Trace(err)
+// GetDDLReorgHandle gets the latest processed DDL reorganize position.
+func (m *Meta) GetDDLReorgHandle(job *model.Job) (startHandle, endHandle, partitionID int64, err error) {
+	startHandle, err = m.txn.HGetInt64(mDDLJobReorgKey, m.reorgJobStartHandle(job.ID))
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	endHandle, err = m.txn.HGetInt64(mDDLJobReorgKey, m.reorgJobEndHandle(job.ID))
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	partitionID, err = m.txn.HGetInt64(mDDLJobReorgKey, m.reorgJobPartitionID(job.ID))
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	// endHandle or partitionID may be 0, because older version TiDB (without table partition) doesn't store them.
+	// update them to table's in this case.
+	if endHandle == 0 || partitionID == 0 {
+		if job.ReorgMeta != nil {
+			endHandle = job.ReorgMeta.EndHandle
+		} else {
+			endHandle = math.MaxInt64
```

Comment:
When will job.ReorgMeta be nil?

Reply:
Here for example: https://github.com/pingcap/tidb/blob/master/ddl/reorg_test.go#L72